### PR TITLE
Adds support for google drive backups on android

### PR DESCRIPTION
**Pull request by ConorOkus**
Closes [#11](https://github.com/photon-sdk/photon-lib/issues/11)

### CODE REVIEW

File: src/GDriveCloudStorage.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  GDrive.setAccessToken(accessToken);
+  GDrive.init();
+  if (!GDrive.isInitialized) {
+    throw new Error('Unable to use GDrive');
+  }
+}
+
+export async function setItem(keyId, value) {
+  const content = Buffer.from(value).toString('base64');
+
```
Comment (tanx) [inline src/GDriveCloudStorage.js:31]:
```suggestion
```

File: src/GDriveCloudStorage.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  return key;
+}
+
+export async function removeItem(keyId) {
+  const fileId = await _getFileId(keyId);
+  if (!fileId) {
+    return null;
+  }
+  await GDrive.files.delete(fileId);
+}
```
Comment (tanx) [inline src/GDriveCloudStorage.js:75]:
```suggestion
}

```

File: src/GDriveCloudStorage.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      name: keyId,
+    },
+    true,
+  );
+}
+
+export async function _getFileId(keyId) {
+  const response = await GDrive.files.list({ spaces: 'appDataFolder', fields: 'nextPageToken, files(id, name)' });
+  const json = await response.json();
+
```
Comment (tanx) [inline src/GDriveCloudStorage.js:46]:
```suggestion
```

File: src/GDriveCloudStorage.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  if (json.files.length === 0) {
+    return null;
+  }
+
```
Comment (tanx) [inline src/GDriveCloudStorage.js:50]:
```suggestion
```

File: src/GDriveCloudStorage.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  const file = json.files.find(file => file.name === keyId);
+
```
Comment (tanx) [inline src/GDriveCloudStorage.js:52]:
```suggestion
```

File: src/GDriveCloudStorage.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  return file ? file.id : null;
+}
+
+export async function getItem(keyId) {
+  const fileId = await _getFileId(keyId);
+
```
Comment (tanx) [inline src/GDriveCloudStorage.js:58]:
```suggestion
```

File: src/GDriveCloudStorage.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+  if (!fileId) {
+    return null;
+  }
+
```
Comment (tanx) [inline src/GDriveCloudStorage.js:62]:
```suggestion
```

File: src/GDriveCloudStorage.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  const response = await GDrive.files.get(fileId, { alt: 'media' });
+  const key = await response.text();
+
```
Comment (tanx) [inline src/GDriveCloudStorage.js:65]:
```suggestion
```

File: src/GDriveCloudStorage.js [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+export async function authenticate(options = {}) {
+  GoogleSignin.configure({
+    scopes: ['https://www.googleapis.com/auth/drive.appdata'],
+    ...options,
+  });
+  await GoogleSignin.hasPlayServices({
+    showPlayServicesUpdateDialog: true,
+  });
+
```
Comment (tanx) [inline src/GDriveCloudStorage.js:12]:
```suggestion
```

File: src/GDriveCloudStorage.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+  try {
+    await GoogleSignin.signInSilently();
+  } catch (error) {
+    if (error.code === statusCodes.SIGN_IN_REQUIRED) {
+      await GoogleSignin.signIn();
+    }
+  }
+
```
Comment (tanx) [inline src/GDriveCloudStorage.js:20]:
```suggestion
```

**Review by tanx — CHANGES_REQUESTED** [review]
@ConorOkus thanks! Gave it a first once over ... just some nits.

Also please squash the diff to the `package-lock.json` from PRs.

And one question... didn't we have `GDriveCloudStorage` module integrated into the `CloudStorage` module for Android? Oh and what about the unit tests?

**Comment by ConorOkus** [comment]
Yeah we did, weird. I think my rebasing messed something up. I've added them back in now anyway!

File: __tests__/unit/cloudstore.android.test.js [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        await expect(CloudStore.removeKeyId({ keyId: 'invalid' })).rejects.toThrow(/not found/);
+        expect(getItemSpy.mock.calls.length).toBe(2);
+
+        getItemSpy.mockRestore();
+        removeItemSpy.mockRestore();
+        getKeySpy.mockRestore();
+      });
+    });
+  });
+});
```
Comment (tanx) [inline __tests__/unit/cloudstore.android.test.js:96]: Nice 👍 Does jest emulate an Android environment if the file is called `*.android.test.js` or how does it work?

**Review by tanx — COMMENTED** [review]
Sweet. Just one question below...

**Comment by tanx** [comment]
Oh, and also the diff to the `package-lock.json` should be removed from the PR. Thanks 🙏

File: __tests__/unit/cloudstore.android.test.js (hunk 11 above)
Comment (ConorOkus) [reply __tests__/unit/cloudstore.android.test.js:99]: No unfortunately not, the naming convention just makes it a bit easier to run in isolation [here](https://github.com/photon-sdk/photon-lib/pull/22/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R16) it's not ideal to have a setup file for running in the android environment but changing the Platform directly in the tests doesn't seem to work so this is a workaround using Haste.

**Comment by ConorOkus** [comment]
Bit confused about removing the diff, I think CI will fail if the package.json and package-lock.json files are not in sync?

**Comment by tanx** [comment]
> Bit confused about removing the diff, I think CI will fail if the package.json and package-lock.json files are not in sync?

Basically just don't commit any changes for the `package-lock.json`. CI shouldn't fail as it will install the newer versions in the `package.json`.

Looks like you deleted the file now?

**Comment by ConorOkus** [comment]
Hmm okay, no I just reverted that change that updates the package-lock.json. Now it's failing in CI?

**Comment by tanx** [comment]
> Hmm okay, no I just reverted that change that updates the package-lock.json. Now it's failing in CI?

You're right. I recently migrated to github actions. Seems like it needs the updated lock file:

npm ERR! cipm can only install packages when your package.json and package-lock.json or npm-shrinkwrap.json are in sync. Please update your lock file with `npm install` before continuing.

Keeping the diff clean worked with Travis. As package maintainer I'm just weary of the security implications of merging and publishing unreviewable dependency updates. If I publish a compromised dependency I'd rather it be my fault.

**Comment by tanx** [comment]
Ok ran the tests locally. Nice solution with the Android unit tests and thank you for the contribution 💯

--> merging 🏖

**Comment by tanx** [comment]
Also updated the `package-lock.json` in https://github.com/photon-sdk/photon-lib/commit/4785b4a79ca818287d9e4468d6f8561005d131c0 and published `v0.8.0` on npm 🎉

**Pull request merged**